frontend/src/metabase/actions/containers/ActionCreatorModal/ActionCreatorModal.unit.spec.tsx

```diff
@@ -102,10 +115,210 @@ describe("actions > containers > ActionCreatorModal", () => {
     const initialRoute = `/model/${MODEL.id}/detail/actions/${action.id}`;
     const { history } = await setup({ initialRoute, action });
 
+    await waitForElementToBeRemoved(() =>
+      screen.queryAllByTestId("loading-spinner"),
+    );
+
     expect(await screen.findByTestId("mock-model-detail")).toBeInTheDocument();
-    expect(screen.queryByTestId("mock-action-creator")).not.toBeInTheDocument();
-    expect(history?.getCurrentLocation().pathname).toBe(
+    expect(screen.queryByTestId("action-creator")).not.toBeInTheDocument();
+    expect(history.getCurrentLocation().pathname).toBe(
       `/model/${MODEL_SLUG}/detail/actions`,
     );
   });
+
+  describe("creating new action", () => {
+    it("does not show custom warning modal when leaving with no changes via SPA navigation", async () => {
+      const initialRoute = `/model/${MODEL.id}/detail/actions`;
+      const actionRoute = `/model/${MODEL.id}/detail/actions/action`;
+      const { history } = await setup({ initialRoute, action: null });
+
+      history.push(actionRoute);
+      await waitForElementToBeRemoved(() =>
+        screen.queryAllByTestId("loading-spinner"),
+      );
+
+      history.goBack();
+
+      expect(
+        screen.queryByText("Changes were not saved"),
+      ).not.toBeInTheDocument();
+      expect(
+        screen.queryByText(
+          "Navigating away from here will cause you to lose any changes you have made.",
+        ),
+      ).not.toBeInTheDocument();
+    });
+
+    it("shows custom warning modal when leaving with unsaved changes via SPA navigation", async () => {
+      const initialRoute = `/model/${MODEL.id}/detail/actions`;
+      const actionRoute = `/model/${MODEL.id}/detail/actions/new`;
+      const { history } = await setup({ initialRoute, action: null });
+
+      history.push(actionRoute);
+      await waitForElementToBeRemoved(() =>
+        screen.queryAllByTestId("loading-spinner"),
+      );
+
+      userEvent.type(screen.getByDisplayValue("New Action"), "a change");
+      userEvent.tab(); // need to click away from the input to trigger the isDirty flag
+
+      history.goBack();
+
+      expect(screen.getByText("Changes were not saved")).toBeInTheDocument();
+      expect(
+        screen.getByText(
+          "Navigating away from here will cause you to lose any changes you have made.",
+        ),
+      ).toBeInTheDocument();
+    });
+
+    it("does not show custom warning modal when saving changes", async () => {
```

Comment:
This scenario might seem redundant, but initially I was getting this problem due to `isDirty` flag not working as expected.